### ISSUE TITLE
Support referring to previous values via automatic variables

### ISSUE THREAD

**Issue by pgebheim** [LiamRiddell/obsidian-solve#6]
I'd love to be able to refer to previous calculations in the same file via a line number reference, similar to the feature of Soulver.

This allows me to break down calculations into multiple lines to self document steps.

**Comment by pgebheim** [comment]
Indeed! I'm looking more for the line-number variables so you can reference a certain line in the file via an automatically variable for a quick calculation.

Can also think of this as result references.

**Comment by LiamRiddell** [comment]
I understand, that's a good idea. However, are you using a plugin to show line numbers in Obsidian? I think it could still be useful to add. Thank you for the feature request!

**Comment by pgebheim** [comment]
Good point with the line numbers...

Maybe there's another way to end up with similar functionality without using line numbers.

Auto-generate a line variable with a short name and include it with the result text?

One weird thing with line numbers is you need to reflow variables as lines are added. Although I used this all the time in soulver it's not necessarily the best ux.

**Comment by LiamRiddell** [comment]
I think the best course of action would be to introduce two modes for Variables. 

1. Viewport Scoped - How it works today and is by far the most performant. 
2. Global Scoped - Search the whole document for variables this would be less performant.

This would allow the user to pick which mode they want based on the average size of documents they use. However, right now, a refresh of variables is later in the roadmap.

**Comment by iamrecursion** [comment]
I love the idea of variables. That would absolutely help with some work I'm doing at the moment. I think that global scoped is the most predictable, but definitely less performant.

**Comment by roaet** [comment]
This is useful to me. At minimum a `$$` variable for "last solve by some line above" would be great.

**Comment by pgebheim** [comment]
> This is useful to me. At minimum a `$$` variable for "last solve by some line above" would be great.

Love this version

**Comment by roaet** [comment]
I think that it is a good thing to be concerned about global scope and
performance at this point. File scope is something easy to remember and
works super well when looking at automatically generated files (like daily)
that use variables.

Having a specific global variable file (or selected imports) would allow
file-variables to work.

**Comment by iamrecursion** [comment]
Yeah I absolutely wouldn't expect this to work in the global scope _between_ notes. Just in the scope of a given file. 

That said, having a specific file like @roaet has mentioned would be pretty cool to define useful constants.

**Comment by roaet** [comment]
There's an issue I added to support import statements. I think import
statements would be a more-better solution than a predefined or
configurable global file.

**Comment by LiamRiddell** [comment]
Thank you for all the feedback everyone. There are some really exciting ideas here. ❤️ 

I think in the short-term, I'll try introducing a reserved variable name of `$$` to refer to the previous result. And for the long term, I like the idea of imports/include statements but that's going to take a lot of work to get it right due to dealing with edge-cases such as circular dependencies, duplicate variable definitions, etc. Additionally, I won't have much time over the next few months due to work/personal commitments. 

Based on all of the above, we'd be looking at that as a post 1.0.0 release of Solve. Which hopefully, will have a bigger community to help support such amazing functionality!

**Comment by iamrecursion** [comment]
`$$` is definitely a great start. Would it be much more work to just have line results just be named `$0`, `$1`, ... and so on? That way we have some way of building complex computations even if we can't define names ourselves.

**Comment by LiamRiddell** [comment]
Hello all, a bit of an update. 

While working on the above I have previous results working great. However, the syntax will be `$prev` because `$$` conflicts with MathJax expressions in Obsidian. This makes me think since MathJax uses `$` extensively we could use `:` prefix instead for example:

1. `100 + 2 - $costOfEggs` becomes `100 + 2 - :costOfEggs`
2. `$prev + 200`  becomes `:prev + 200`
3. `$myVariable = 100` becomes `:myVariable = 100`

@iamrecursion As for line number support, I need to do some code refactoring before I tackle the bigger work that would include line numbers and includes together.

**Comment by iamrecursion** [comment]
I've no issue with `:` as the prefix! 

To clarify, I wasn't actually suggesting actual line numbers in this case! Just having each solve line be sequentially numbered starting at `:1`. I imagine this results in the same complexity, however!

**Comment by Stakker** [comment]
Numi (https://numi.app) has the "prev" feature, and I'm really missing it in Solve, so I wholeheartedly +1 it :-)

Numi also has other similar keywords like "sum", which is very useful – sums up all the rows above until an empty line. (Also "average" works the same way.)

**Comment by LiamRiddell** [comment]
Hey all, this whole thread is now supported, and it's a fun one to close since this is where the `:` variable prefix was originally decided.

The engine rewrite (Solve 2.0, currently in beta) adds cross-line data access. `prev` refers to the immediately preceding line:

```
100 + 2
prev - 50
```
gives `102`, then `52`.

`line1`, `line2`, and so on refer to any specific line by number, which covers the "result references" idea from earlier in the thread without needing manual reflowing:

```
:costOfEggs = 5
100 + 2 - line1
```
gives `5`, then `97`.

And `sum`/`total`/`average` over a range or "above" also made it in, `total above` sums everything back to the top of the document or the last blank line/heading, and `sum(line 1 : line 4)` sums an explicit range.

`prev` and `line<N>` ended up as bare words rather than `:prev`/`:24`, matching Numi's own naming, since they don't collide with anything else in the grammar the way `sum`/`total` would as bare words (those needed the "above" phrase specifically to stay safe next to a variable named `sum`).

This will ship as part of Solve 2.0 (Beta). Closing this out, and thanks to everyone here, this thread genuinely shaped the design.
